### PR TITLE
Remove admin process in favor of daemon in VNet

### PR DESCRIPTION
**Pull request by ravicious**
This PR is an update to [RFD 163](https://github.com/gravitational/teleport/pull/37834). It suggests adding a launch daemon that handles merely TUN device creation and DNS configuration, while everything else is handled by an unprivileged process.

## What we planned so far

The idea was to run a daemon as root and do everything through this daemon. This seemed to be easier: we'd no longer have to pass fd of the TUN device from a one process to another. There would be no need for IPC as the daemon would handle everything. It also seemed more portable as on Windows we'd have to solve this issue with something else than fds.

There are some downsides to this:

* We run a lot of code as root, even though it's needed only for creating a TUN device and setting up DNS.
* As launch daemons start before the actual user logs in, we'd have to somehow detect login & logout to start & stop VNet for the given user.
* Almost every call to the daemon would need to inspect permissions somehow, to make sure that the user that wants to modify the given instance of VNet is the one in whose context VNet is ran.

## How it can be done safely instead

Instead of doing everything through the launch daemon, we can limit the launch daemon only to creating TUN devices and configuring DNS. Actual proxying would be done through a launch agent, which runs as user and automatically gets started and stopped by launchd on login/logout.

Let's take at the list of potential problems an how we can solve them:

* IPC: since we want to use a daemon, we have to IPC with it anyway to be able to start and stop VNet and get recent connections. This can be done by running a gRPC service over a Unix socket at `/var/run/teleport-vnet.socket`, chmodded to 666. 
* Passing TUN device and portability: [grpcfd](https://pkg.go.dev/github.com/edwarnicke/grpcfd) is a package which enables passing fds over gRPC when using Unix sockets. On Windows, [the tun package accepts a GUID when creating a TUN device](https://github.com/WireGuard/wireguard-go/blob/12269c2761734b15625017d8565745096325392f/tun/tun_windows.go#L66-L68) and reuses a Wintun device if the GUID matches an existing device. So hopefully we should be able to share the TUN device from the root process to an unprivileged one no matter the platform.
* Detecting login and logout: done out of the box by launchd.
* We run as root only what's necessary. We don't have to worry about permissions on the proxying side because everything is ran as the given OS user.
   * If there's any auxiliary state we need to save, e.g. whether to start VNet automatically on system start, we can easily save this data straight to the tsh home dir.

## Why we should keep VNet tied to tsh/Connect lifecycle anyway

After all, I still think in v1 the lifecycle of VNet should be tied to the underlying UI.

Without any UI, it's impossible for the user to tell whether VNet is running or not until they try to access a resource. Even then, a failure to access a
resource might stem not only from VNet not running, but it also might be a problem with access or simply a communication problem with the cluster. Without any UI, we have no way of communicating that to the user.

Even if we decide to screw this and just ship it without the UI, keeping VNet running without tsh/Connect still adds a considerable amount of work that I think is not necessary for v1.

## What we end up with now (in this PR)

This PR updates the RFD with the consequences of the following:

* We have a launch daemon running as root, added to login items through `SMAppService` (macOS 13.0+).
* The daemon starts on system login and exposes a gRPC service at `/var/run/teleport-vnet.socket`.
* tsh talks with the daemon whenever it needs to create a TUN device or configure DNS.
* Everything else happens outside of the daemon process.

TODO:

- [ ] Add proto definition for the gRPC service if we decide it's something we want to go with.
- [x] PoC of passing TUN device fd with grpcfd.

### CODE REVIEW

**Review by zmb3 — APPROVED** [review]

**Comment by ravicious** [comment]
I was able to create a TUN device in a root process, then pass the fd over with grpcfd to an unprivileged process and then successfully call `tun.CreateTUNFromFile`.

**Comment by ravicious** [comment]
The PoC of an SMAppService daemon and sending fd over gRPC is at https://github.com/gravitational/vnet-daemon-poc. Though if we plan to stick with osascript for Q1, then none of that matters until we implement all the other stuff first.

File: rfd/0163-vnet.md [hunk 1]
```diff
@@ -426,39 +426,51 @@ nameserver 100.127.100.127
...
+plist lives in the app bundle, it is automatically removed when the user removes tsh.app or Teleport
+Connect.app.
 
-The admin process will be started in the background by `tsh vnet` when it first
-starts, or by Connect when VNet is first turned on (usually when it first opens
-after the first time VNet is enabled).
-It will watch or poll the contents of the `TELEPORT_HOME` directory for
-currently active Teleport profiles.
+`SMAppService` requires macOS 13.0+. Since it only works with signed packages, in development we can
+substitute it with an osascript wrapper that runs the same command as root.
```
Comment (nklaassen) [inline rfd/0163-vnet.md:443]: How much would introducing codesigning change our current development/deployment/installation process for Connect? Is it worth it in a v1 if we could get away without SMAppService by creating a plist file for a launchd launch daemon  https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html?

Comment (ravicious) [reply rfd/0163-vnet.md:443]: As discussed yesterday, we already do codesigning and we will eventually use SMAppService.

We could use `launchctl` directly, but now that I wrote a PoC of SMAppService, it doesn't really save us any time and introduces additional problems. For example, SMAppService automatically cleans up the daemon when trashing `.app` while with launchctl the user would have to manually remember to remove the daemon plist ([AFAIK](https://forums.developer.apple.com/forums/thread/66821?answerId=245716022#245716022)).

File: rfd/0163-vnet.md [hunk 2]
```diff
@@ -426,39 +426,51 @@ nameserver 100.127.100.127
...
 
-The process will create clients for all currently logged-in profiles and all of
+The daemon allows other processes to talk to it through a gRPC service available over a Unix socket
+at `/var/run/teleport-vnet.socket`. Once the daemon establishes a listener on that socket, it chmods
+the socket to 666 to allow unprivileged processes to use the gRPC service.
```
[the remark this answers is not in the document]
Comment (ravicious) [reply rfd/0163-vnet.md:447]: As the plist has to be embedded within the app bundle, we cannot dynamically add a user in there.

The API for `SMAppService` accepts only the plist filename. I couldn't find a straightforward way to somehow express "Tell the daemon that this was the user who added the plist".

We could probably do some chicanery and detect the user on the first daemon start. I'll leave a publicly available socket for now and then we'll see what people say during the review.

Technically, we could easily make it available at first and then chmod it to a specific user on the first connection made to the service.

File: rfd/0163-vnet.md [hunk 3]
```diff
@@ -426,39 +426,51 @@ nameserver 100.127.100.127
...
 in mind, but we can implement this in a similar way for Windows and Linux.
 
-#### Admin process
+#### Daemon
 
-All system configuration must run as a privileged user (root).
+All system configuration must run as a privileged user (root). To achieve that, VNet starts a daemon
+with `tsh vnet daemon` that runs as root, a Launch Daemon in macOS terms. The daemon handles
+creating a TUN device (and passing back its file descriptor or [GUID on
+Windows](https://github.com/WireGuard/wireguard-go/blob/12269c2761734b15625017d8565745096325392f/tun/tun_windows.go#L66-L68)).
```
[the remark this answers is not in the document]
Comment (ravicious) [reply rfd/0163-vnet.md:434]: We resolved the issue with lifetimes by limiting the root daemon to allowing only one TUN device running at a time.

**Pull request merged**